byzcoin/transaction.go

```diff
@@ -206,38 +207,60 @@ func (instr Instruction) DeriveID(what string) InstanceID {
 // Action returns the action that the user wants to do with this
 // instruction.
 func (instr Instruction) Action() string {
+	contractID := instr.ContractID()
 	a := "invalid"
 	switch instr.GetType() {
 	case SpawnType:
-		a = "spawn:" + instr.Spawn.ContractID
+		a = "spawn:" + contractID
 	case InvokeType:
-		a = "invoke:" + instr.Invoke.ContractID + "." + instr.Invoke.Command
+		a = "invoke:" + contractID + "." + instr.Invoke.Command
 	case DeleteType:
-		a = "delete:" + instr.Delete.ContractID
+		a = "delete:" + contractID
 	}
 	return a
 }
 
-// String returns a human readable form of the instruction.
-func (instr Instruction) String() string {
-	var out string
-	out += fmt.Sprintf("instr: %x\n", instr.Hash())
-	out += fmt.Sprintf("\tinstID: %v\n", instr.InstanceID)
-	out += fmt.Sprintf("\taction: %s\n", instr.Action())
-	out += fmt.Sprintf("\tidentities: %v\n", instr.SignerIdentities)
-	out += fmt.Sprintf("\tcounters: %v\n", instr.SignerCounter)
-	out += fmt.Sprintf("\tsignatures: %d\n", len(instr.Signatures))
+// ContractID returns the ContractID that the user specified
+func (instr Instruction) ContractID() string {
+	a := "invalid"
 	switch instr.GetType() {
 	case SpawnType:
-		out += fmt.Sprintf("Spawn:\t%s\n\tArgs:%s\n", instr.Spawn.ContractID,
-			strings.Join(instr.Spawn.Args.Names(), " - "))
+		a = instr.Spawn.ContractID
 	case InvokeType:
-		out += fmt.Sprintf("Invoke:\t%s\n\tArgs:%s\n", instr.Invoke.ContractID,
-			strings.Join(instr.Invoke.Args.Names(), " - "))
+		a = instr.Invoke.ContractID
 	case DeleteType:
-		out += fmt.Sprintf("Delete:\t%s\n", instr.Delete.ContractID)
+		a = instr.Delete.ContractID
 	}
-	return out
+	return a
+}
+
+// String returns a human readable form of the instruction.
+func (instr Instruction) String() string {
+	contractFn, ok := ContractsFn[instr.ContractID()]
+	var methodStr string
+	if !ok {
+		methodStr = "error in getting constructor: " + instr.ContractID()
+	} else {
+		contract, err := contractFn(nil)
+		if err != nil {
+			methodStr = "error in getting contract"
+		} else {
+			methodStr = contract.PrintMethod(instr)
+		}
+	}
+
+	var out strings.Builder
+	out.WriteString("- instruction:\n")
+	fmt.Fprintf(&out, "-- hash: %x\n", instr.Hash())
+	fmt.Fprintf(&out, "-- instID: %v\n", instr.InstanceID)
+	fmt.Fprintf(&out, "-- action: %s\n", instr.Action())
+	fmt.Fprintf(&out, "-- identities: %v\n", instr.SignerIdentities)
+	fmt.Fprintf(&out, "-- counters: %v\n", instr.SignerCounter)
+	fmt.Fprintf(&out, "-- signatures: %d\n", len(instr.Signatures))
+	out.WriteString(
+		regexp.MustCompile(`(?m)^(.+)$`).ReplaceAllString(methodStr, "-$1"))
```

Comment:
Regexp compilation should happen in the package scope, so that they happen once during init, instead of on every execution of String(). (Also the other call to MustCompile above.)

A comment on this regexp would be helpful too.